src/diffusers/pipelines/controlnet_xs/pipeline_controlnet_xs.py

```diff
@@ -139,6 +185,85 @@ def __init__(
         )
         self.register_to_config(requires_safety_checker=requires_safety_checker)
 
+    @classmethod
```

Comment:
you should not need to define `from_pretrained` and `save_pretrained`, instead define your `__init__` as

```
def __init__(
        self,
        ...
        unet: UNet2DConditionModel,
        controlnet: ControlNetXSAddon,
        ...
    ):
        super().__init__()
        self.controlnet = ControlNetXSModel(unet, controlnet)
  ```
and ` time_embedding_mix` should be a config of `ControlNetXSAddon`

this API should still work
```python
controlnet_xs_addon = ControlNetXSAddon.from_pretrained(...)
pipe = StableDiffusionControlNetXSPipeline.from_pretrained(..., controlnet=controlnet_xs_addon)
```

Reply:
I believe we need to change `from_pretrained` and `save_pretrained`.

The core issue is
    (1) We want the loading api to be `pipe.from_pretrained('sd_path', controlnet_addon)` which requires `controlnet_addon` to be in the init signature, which requires it to be registered to config.
    (2) We also want cpu offloading to work, which requires `controlnet` (not `controlnet_addon`) to be registered to config and therefore be in the init signature. This is because cpu offloading moves models to gpu when their foward is called, but only does this for config-registered modules. And only `controlnet` has a forward, as `controlnet_addon` is not a full, independent model.

I think the best solution is to have `controlnet` in the init signature, which makes (2) work, and change  `from_pretrained`/`save_pretrained` to also make (1) work.

What do you think?

Reply:
would this `__init__` work? 

```
def __init__(
        self,
        ...
        unet: UNet2DConditionModel,
        controlnet: ControlNetXSAddon,
        ...
    ):
        super().__init__()
        controlnet = ControlNetXSModel(unet, controlnet)
        self.register_modules(
            unet=unet,
            controlnet=controlnet,
            scheduler=scheduler,
            ...
        )
```

Reply:
also I'm not sure if the cpu offloading will work here for unet and controlnet because it isn't sequential

Reply:
> would this `__init__` work?
> 
> ```
> def __init__(
>         self,
>         ...
>         unet: UNet2DConditionModel,
>         controlnet: ControlNetXSAddon,
>         ...
>     ):
>         super().__init__()
>         controlnet = ControlNetXSModel(unet, controlnet)
>         self.register_modules(
>             unet=unet,
>             controlnet=controlnet,
>             scheduler=scheduler,
>             ...
>         )
> ```

Unfortunately no, because the registered modules and the init signature need to be the same. I've created a minimal example here: https://github.com/UmerHA/One-Off-Notebooks/blob/main/controlnet_xs_pipe_init.ipynb

Reply:
> also I'm not sure if the cpu offloading will work here for unet and controlnet because it isn't sequential

I think we need to be clear about the involved objects. We have `unet`, `controlnet_xs_addon` and `controlnet_xs`.

The `unet` and `controlnet_xs_addon` are combined to a `controlnet_xs`. This differs from the usual controlnet, where we have 2 objects (`unet`, `controlnet`) that can be run independently. In controlnet_xs, the `unet` and `controlnet_xs_addon` get interwoven, which is why we need a new `controlnet_xs` object.

So cpu offloading will work,  because we don't have a `unet` object anymore, only a `controlnet_xs` object.

Reply:
hey! thanks for explaining and making the notebook! 
let's do 
```python
def __init__(
        self,
        ...
        unet: UNet2DConditionModel,
        controlnet: ControlNetXSAddon,
        ...
    ):
        super().__init__()
        unet = ControlNetXSModel(unet, controlnet)
        self.register_modules(
            unet=unet,
            controlnet=controlnet,
            scheduler=scheduler,
            ...
        )
```
Basically our `unet` here will be `unet` with `controlnet` injected, we should also consider changing the class name `ControlNetXSModel` as well, maybe `UNetControlNetXSModel`?
it is similar to what's done for these video pipelines https://github.com/huggingface/diffusers/blob/6c5f0de7139ee538536478146ff6b3ac5c4ed055/src/diffusers/pipelines/animatediff/pipeline_animatediff.py#L148

I updated your notebook too:
https://colab.research.google.com/drive/1Jzyo3z_fKrAdbh7CfmoJQRDOwsZQTqaY?usp=sharing

Reply:
Hey @yiyixuxu - done :)

I've created a `UNetControlNetXSModel` analogous to `UNetMotionModel`, and created tests as well.

Reply:
@yiyixuxu - hey! pinging you again as this has been stale for a week :)